lib/hanami/action.rb

```diff
@@ -435,36 +430,38 @@ def self.new(configuration:, **args)
       end
     end
 
-    module InstanceMethods
-      # Implements the Rack/Hanami::Action protocol
-      #
-      # @since 0.1.0
-      # @api private
-      def call(env)
-        request  = nil
-        response = nil
-        halted   = catch :halt do
-          begin
-            params   = self.class.params_class.new(env)
-            request  = Hanami::Action::Request.new(env, params)
-            response = Hanami::Action::Response.new(action: self.class.name, configuration: configuration, content_type: Mime.calculate_content_type_with_charset(configuration, request, accepted_mime_types), env: env, header: configuration.default_headers)
-            _run_before_callbacks(request, response)
-            super(request, response)
-            _run_after_callbacks(request, response)
-          rescue => exception
-            _handle_exception(request, response, exception)
-          end
+    # Implements the Rack/Hanami::Action protocol
+    #
+    # @since 0.1.0
+    # @api private
+    def call(env)
+      request  = nil
+      response = nil
+
+      halted = catch :halt do
+        begin
+          params   = self.class.params_class.new(env)
+          request  = Hanami::Action::Request.new(env, params)
+          response = Hanami::Action::Response.new(action: self.class.name, configuration: configuration, content_type: Mime.calculate_content_type_with_charset(configuration, request, accepted_mime_types), env: env, header: configuration.default_headers)
+          _run_before_callbacks(request, response)
+          handle(request, response)
+          _run_after_callbacks(request, response)
+        rescue => exception
+          _handle_exception(request, response, exception)
         end
-
-        finish(request, response, halted)
       end
+
+      finish(request, response, halted)
     end
 
     def initialize(**)
     end
 
     protected
 
+    def handle(request, response)
```

Comment:
@jodosha since it's possible for an `Action` subclass to do all its work via callbacks (i.e. without any custom `#handle`), I figured it would be best for the default implementation simply to be a no-op, rather than something that raises a "not implemented" exception. If we had the default implementation raising an exception, we might end up requiring our users to define _empty_ `#handle` methods, which seems undesirable.

As for how the `#handle` method should be protected (or not), I think that having it as protected in the first place is the right thing to do, since this isn't meant to be part of an action object's public API.

And as for what happens when a subclass overrides `#handle`,  you're right that it'll become public unless it's explicit re-set as protected or private, but I don't think that's any reason to change the default visibility?

(In real terms, if `#handle` ends up being public for action subclasses, it's hardly the worst thing, since it'll be clear from every piece of documentation that `#call` is the interface to use)

If we _really_ wanted to be strict about this, I suppose we could do something like this:

```ruby
class Base
  def self.method_added(method_name)
    protected method_name if method_name == :handle
  end

  protected

  def handle
  end
end

b = Base.new
b.handle

class Sub < Base
  def handle
  end
end

s = Sub.new
s.handle
# NoMethodError: protected method `handle' called for #<Sub:0x00007fdcaa2db098>
```

But IMO it's not worth it. It's messing with other people's classes in a way they might not expect. And Ruby's visibility rules have always been more _suggestions_ than hard and fast rules, anyway.

Thoughts?

Reply:
@timriley I don't want to enforce any visibility rule on final devs' objects. What I'm saying is: why not make it public in the first place?

Reply:
Fine. Let's leave it as it, but let's revisit later if needed.